buildkite/resource_team_member.go

```diff
@@ -3,174 +3,208 @@ package buildkite
+				},
 			},
-			"user_id": &schema.Schema{
-				Required: true,
-				Type:     schema.TypeString,
+			"team_id": resource_schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "The GraphQL ID of the team to add/remove the user to/from.",
 			},
-			"uuid": &schema.Schema{
-				Computed: true,
-				Type:     schema.TypeString,
+			"user_id": resource_schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "The GraphQL ID of the user to add/remove.",
+			},
+			"role": resource_schema.StringAttribute{
+				Optional:            true,
+				MarkdownDescription: "The role for the user. Either MEMBER or MAINTAINER.",
+				Validators: []validator.String{
+					stringvalidator.OneOf("MEMBER", "MAINTAINER"),
+				},
 			},
 		},
 	}
 }
 
-// CreateTeamMember adds a user to a Buildkite team
-func CreateTeamMember(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	originalRole := d.Get("role").(string)
-	client := m.(*Client)
-
-	var mutation struct {
-		TeamMemberCreate struct {
-			TeamMemberEdge struct {
-				Node TeamMemberNode
-			}
-		} `graphql:"teamMemberCreate(input: {teamID: $teamId, userID: $userId})"`
-	}
+func (tm *teamMemberResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan, state teamMemberResourceModel
 
-	vars := map[string]interface{}{
-		"teamId": graphql.ID(d.Get("team_id").(string)),
-		"userId": graphql.ID(d.Get("user_id").(string)),
-	}
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
 
-	err := client.graphql.Mutate(context.Background(), &mutation, vars)
-	if err != nil {
-		return diag.FromErr(err)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	updateTeamMember(d, &mutation.TeamMemberCreate.TeamMemberEdge.Node)
+	log.Printf("Creating team member into team %s ...", plan.TeamId.ValueString())
+	apiResponse, err := createTeamMember(
+		tm.client.genqlient,
+		plan.TeamId.ValueString(),
+		plan.UserId.ValueString(),
+		TeamMemberRole(*plan.Role.ValueStringPointer()),
+	)
 
-	// theres a bug in teamMemberCreate that always sets role to MEMBER
-	// so if using MAINTAINER, make a separate call to change the role if necessary
-	if originalRole == "MAINTAINER" {
-		d.Set("role", "MAINTAINER")
-		UpdateTeamMember(ctx, d, m)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to create team member",
+			fmt.Sprintf("Unable to create team member: %s", err.Error()),
+		)
+		return
 	}
 
-	return nil
+	// Update state with values from API response/plan
+	state.Id = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Id)
+	state.Uuid = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Uuid)
+	state.TeamId = plan.TeamId
+	state.UserId = plan.UserId
+	// The role of the user will by default be "MEMBER" if none was entered in the plan
+	state.Role = types.StringValue(string(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Role))
```

Comment:
Is this still true? There used to be a bug in the API but I think its been resolved. If not specified, the default will be set to the default configured on the team. So this could possibly modify state and cause a subsequent apply to generate a diff which is something we don't want. We should add a test for that edge case. Happy to pair on that with you

Reply:
Per the chat - I think leaving the resource as is from the older SDKv2 impl (with role always required) would keep this for a future improvement/update 
Happy for other thoughts though 👍